### ISSUE TITLE
Computer player = COMPUTER3?

### ISSUE THREAD

**Issue by peyre** [raceintospace/raceintospace#323]
This is an oddball, but a bit obscure.  Like the flag says, this isn't at all high-priority, but since I spotted it I thought I'd mention it.

I just finished a game, got whipped by the Soviets flying Voskhod.  I left the computer player as "KOROLOV", but the endgame identifies it as "COMPUTER3".

![Korolov](https://user-images.githubusercontent.com/3118014/74306934-d2d5c400-4d18-11ea-9926-da557a4b3c47.png)
![Computer3](https://user-images.githubusercontent.com/3118014/74306938-d701e180-4d18-11ea-9c67-3cc132fef02d.png)

**Comment by rnyoakum** [comment]
Hmm... I did a quick search, and that screen appears to be generated by the `DrawStatistics(char Win)` function in aipur.cpp. The "Winning Director" code is found on lines 79-90, and the "Losing Director" code on lines 92-104. In the case of a computer player, RIS reports the director as being a computer, followed by the numerical code of the strategy the AI was using.

It's simple enough to replace "COMPUTER(X)" with the name of the AI's director, but I do think it cool that it reports the strategy used. That's something I would like to preserve. Even better would be to explain the strategy, because "COMPUTER3" is exactly the sort of comment that even avid players find unhelpful.

**Comment by peyre** [comment]
One thing confuses me about that: the computer won with a Voskhod mission, and I don't think I recall seeing it fly anything else.  Shouldn't it have said COMPUTER0 or COMPUTER1?  Or is there a number for each of the individual types of strategy: Voskhod HIstorical, Voskhod EOR, Voskhod LOR, Soyuz Historical, etc.?

Showing the strategy used is a good idea though.  Maybe below the WINNING DIRECTOR: KOROLOV line it could show something like PRIMARY STRATEGY: SOYUZ or STRATEGY USED: SOYUZ HISTORICAL.

**Comment by peyre** [comment]
Well this is confusing.  The computer just beat me as the US, flying direct ascent, and it's identified as COMPUTER1.

![Computer1](https://user-images.githubusercontent.com/3118014/75631370-31b78c00-5ba7-11ea-882f-5408fdf242a9.png)
![Jupiter](https://user-images.githubusercontent.com/3118014/75631371-3419e600-5ba7-11ea-8620-f362e9ab321f.png)

**Comment by peyre** [comment]
So, I've done some experimenting with it today, and it seems that:
1 = Direct Ascent
2 = Apollo/Soyuz
3 = Gemini/Voskhod
Very interesting that the minishuttles seem to have been skipped.  Maybe the computer is programmed to not consider flying them.

**Comment by peyre** [comment]
Ok, I've updated aipur.cpp to specify the computer's strategy.  In the event that it returns a value other than 1, 2, or 3, it will say "UNKNOWN (x)" so we'll be able to tell what strategy x is: minishuttle, C-Kicker, or whatever.

**Comment by rnyoakum** [comment]
Based on the images you captured - https://github.com/raceintospace/raceintospace/issues/331#issuecomment-598858518 and https://github.com/raceintospace/raceintospace/issues/331#issuecomment-598859343 - I think it's clear the strategy is not a Spacecraft specification. The image clearly shows the flight was Soyuz V, not the LK-700. I'm reopening this and going to take a shot at it.

**Comment by peyre** [comment]
That makes sense.  Thanks Ryan.  I wonder if the strategy variable indicates direct ascent vs. historical vs. EOR vs. LOR vs. Soyuz LL.

**Comment by peyre** [comment]
And again, the computer beat me with direct ascent and shows as COMPUTER3.  I wonder if the number refers to level of risk-taking assigned to the AI, or what.

![LK700](https://user-images.githubusercontent.com/3118014/76703510-68090700-668f-11ea-8ac0-ca840b9bef0d.png)
![Computer3](https://user-images.githubusercontent.com/3118014/76703512-6b03f780-668f-11ea-9ac1-f39296aa08de.png)

**Comment by peyre** [comment]
And in my game today, the computer was flying Soyuz and still showed as COMPUTER3.

**Comment by rnyoakum** [comment]
Are you certain that you're running the latest build of the game? From the looks of the code, it shouldn't be outputting COMPUTERX as the name (unless, well, that was the name you gave the director).

**Comment by peyre** [comment]
I wondered if you'd notice that.  I'm running a slightly older build.  We're up at a rental house in Truckee (and kind of snowed in here) so I'm running it on my laptop.  I downloaded the source and tried to compile it, but it errors out on prefs.cpp; I think something Petr did with it recently might be causing it to barf.

make[2]: *** [src/game/CMakeFiles/game_test.dir/build.make:752: src/game/CMakeFiles/game_test.dir/prefs.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:351: src/game/CMakeFiles/game_test.dir/all] Error 2
make: *** [Makefile:163: all] Error 2

**Comment by rnyoakum** [comment]
Yeah, I noticed that. Petr's last commit accidentally removed a variable declaration that was needed, so it won't compile. I'm planning to fix that tonight.

**Comment by peyre** [comment]
Oh, awesome!  I'll keep an eye on it and download when I see the next commit.

**Comment by peyre** [comment]
Here's another.  I see you changed "COMPUTER" to "(AI)".  Makes sense to me.  The computer flew (and won with) LK-700, but its strategy shows as VOSKHOD.  Now the computer did fly Voskhod earlier in the game; maybe that's what happened, that it flew Voskhod but didn't change its strategy numeral when it switched to direct ascent?

![LK700](https://user-images.githubusercontent.com/3118014/76872754-11840000-682a-11ea-8a6e-b855e5a4f531.png)
![AI](https://user-images.githubusercontent.com/3118014/76872771-16e14a80-682a-11ea-9ba4-240 68d95f63a.png)

**Comment by rnyoakum** [comment]
Yeah, I changed COMPUTER to (AI) so it would append to a computer player's name without overflowing the space, but I haven't gotten the name fixed yet. That should be coming in the next day or two; I was hoping to test to make sure the fix works.

As for the strategy, there looks to be three different functions the AI chooses between to define its approach. They aren't documented, and are called Strategy_One(), Strategy_Two(), and Strategy_Thr(). Not the most illuminating names. In order to understand the differences between them, I'm going to have to read over all their code and figure out what, exactly, they are doing and why. That's going to take longer than the name fix.

**Comment by peyre** [comment]
Very good!  Thanks for looking into it.  We'll see how it comes out.

**Comment by rnyoakum** [comment]
Yeah, I'm a little concerned over that, but " (AI)" is pretty short so why not try it? It works for the default director names, but if you think that's still a problem I'll take it out. How many players change the default AI names? Regardless, I finally realized I had a save near the end of the game and so could test my fix.

This should be resolved by commit 300b8167f84f0590bb1a85f971b9875caf8afa8d.

**Comment by peyre** [comment]
Sure, if it works fine, I'd rather leave it on the same line.  I was just pointing out a potential concern that occurred to me.

So this is fixed?  I'll have to try it out!

**Comment by peyre** [comment]
So, I thought maybe I could save you some time and effort by asking Michael McCarty if he remembered what that variable (those variables) represent(s).  I mentioned that it doesn't correspond to the AI using Voskhod, Soyuz, or direct ascent as its means to do the Moon landing.  His reply came in this morning:

> I think the strategies are mostly on the path used.  It would be interesting to use Graphviz to plot each one and compare them visually.

Does that make any sense to you, or did he just tell me he thinks it corresponds to spacecraft type?

**Comment by peyre** [comment]
Michael just clarified for us: 

> More the path rather than the craft.  Each of the styles try to plan missions on each pad based on a step that has been passed.

**Comment by peyre** [comment]
So, found this: the strategy is set randomly in aimast.cpp, lines 74-84.  Good to know.  And judging by what Michael had to say, I guess . . . well I don't know.  But maybe today I'll edit aipur.cpp so it no longer tells you the computer's strategy was LK-700 when you just watched it go to the Moon on a Soyuz.  Maybe I'll just have it say the computer used strategy 1, 2, or 3, without trying to ascribe a specific meaning to the number.

**Comment by peyre** [comment]
Ok, I've fixed it so now it shows AI STRATEGY: 1 (or 2 or 3).

![Statistics](https://user-images.githubusercontent.com/3118014/82125849-9a051b00-975d-11ea-9f5c-2f779998ef4a.png)

**Comment by peyre** [comment]
I had a closer look at aimast.cpp today.  One of the comments says "strategy 0=none 1=one 2=two" and further down where it handles AI_BEGINNING_STRATEGY, it looks like 0 charges ahead to the orbital EVA if the suits are ready and does the suborbital otherwise, while 1 and 2 take a more step-by-step approach, but 2 is less cautious.  So it kind of looks like 0 is random, 1 is (sort of) cautious, and 2 is more daring.

**Comment by peyre** [comment]
It looks like early on the code changes 0, 1, and 2 to 1, 2, and 3.  Comparing Strategy_Two vs. Strategy_Three in aimis.cpp, it looks there as though Two may be more daring in the mid stage of the game.  Strategy_One and Strategy_Three look almost identical in aimis.cpp.

Also, there are various other strategies used, including AIStrategy[AI_BEGINNING_STRATEGY] mentioned above, AIStrategy[AI_LARGER_ROCKET_STRATEGY], and so forth.  I'm not sure if they correspond to the AIStrategy[AI_STRATEGY] (which is randomly determined in aimast.cpp) or not.

**Comment by peyre** [comment]
Incidentally, I've been keeping a log of my games, keeping track of the computer's strategy number, what they were flying, and anything about their strategy that strikes me.  Maybe I'll see a pattern eventually.

**Comment by peyre** [comment]
Looking more closely, I see that much of the difference between strategies seems to be that one will try a Lunar Orbital LM Test sometimes while the other will only do LM Orbitals.  There really don't look to be too many differences between strategies.  Maybe that's to be expected, or maybe they're subtle differences that make a big difference in the course of the game.

**Comment by peyre** [comment]
Comparing them this morning, I see that there are several thresholds here, such as if DM Safety ≥80%, it proceeds with a Lunar Orbital or Landing.  It's tempting to differentiate them a bit by tweaking one of them to make that AI strategy a bit more cautious or risk-taking.  Of course, that's not the sort of change you make lightly; it would need a lot of testing.

**Comment by peyre** [comment]
I think I was right back in 2020: this one can be closed.  'Course we can always reopen it if something comes up.
